server/Dockerfile

```diff
@@ -29,9 +29,13 @@ RUN --mount=type=bind,source=./03-python-base,target=./03-python-base ./03-pytho
 
 FROM python-base-internal as python-dev-base
 RUN --mount=type=bind,source=./04-python-dev-base,target=./04-python-dev-base ./04-python-dev-base/install.sh
+ENV VIRTUAL_ENV='/opt/build-venv' \
+   PATH="/opt/build-venv/bin:${PATH}"
 
 FROM python-base-internal as server-base
 RUN --mount=type=bind,source=./04-server-base,target=./04-server-base ./04-server-base/install.sh
+ENV VIRTUAL_ENV='/opt/deephaven-venv' \
```

Comment:
PATH can be given directories that don't exist yet - you could set this before either 04- operation, and then the newly created bins will be available right away, only need one `ENV` instead of two that way.

that said, since later commands rely on the PATH that both of these have appended to (with multiple `python`s on the PATH), would it make more sense to overwrite PATH to remove build-venv, or just use activate in each script to specify which one is expected, etc? 

It isn't obvious to me why build-venv is added to the path after 04-python-dev-base, since 04-server-base still references `python3`, but scripts 05-* all just reference `python` - is this deliberate? python just points to python3 I believe?

Reply:
04-server-base does not derive from 04-python-dev-base (the linear looking nature of `server/Dockerfile` may lead to that confusion).

I'm not a fan of setting `PATH` before a `RUN` command in general - but also, in this specific case it would cause these two lines to execute with a different command which is very confusing IMO:

```shell
python3 -m venv /opt/deephaven-venv
python3 -m pip install -q --no-cache-dir --upgrade pip
```

I'm preferring to have the downstream commands reference `python` instead of `python3` (which does exist in the venv) partially b/c I think it's more idiomatic shell scripting when `python` is available, but also, if a venv is not setup, it will *fail* (`python` binary/link is *not* created as part of the apt package `python3.7`).

On the further downstream (deephaven-core) jpy side of things, we have something like:

```shell
python3.7 setup.py bdist_wheel
```

which is *not* setup as part the venv; so we'll need to change that invocation to `python3` or `python`, and my preference is to assume virtual env and use `python`.

Reply:
Note:

the venv activate script does two essential things (`python3 -m venv /tmp/testing` on my local box):

```
VIRTUAL_ENV="/tmp/testing"
export VIRTUAL_ENV

PATH="$VIRTUAL_ENV/bin:$PATH"
export PATH
```

some python tooling prefers to check `VIRTUAL_ENV` as a more specific indication of working in a virtual envirnoment, rather than searching the `PATH` for python.

Reply:
> it would cause these two lines to execute with a different command which is very confusing IMO

Those two lines would be `python3`, `python` respectively, at least going by their current conventions. We do expect different scripts (even those run in the same image) which use `python3` vs `python` to actually run different binaries, perhaps a comment would be clearer? Relying on "this bin is fully qualified" vs "this isn't" is what I'm drawing attention to, both in the same script, and across scripts. 

Your note about how `python` will fail in the default docker image applies here too.

Again, not a terribly strong preference, it just seems incongruous that `04-*` uses one convention to call the venv'd python (not python3), and the `05-*` uses a different one, the reader is expected to have noticed the ENV that runs between them, rather than taking the scripts at face value.

Reply:
I agree - we can follow the same standard here; we'll just need to know that the venv creator sourcing the activate script is only applicable for itself - and everything else inherits from the ENV.